server/sonar-process/src/main/java/org/sonar/process/jmx/JmxConnection.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.process.jmx;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MemoryMXBean;
+import java.lang.management.ThreadMXBean;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import javax.annotation.CheckForNull;
+import javax.management.JMX;
+import javax.management.MBeanServerConnection;
+import javax.management.ObjectName;
+import javax.management.remote.JMXConnector;
+
+public class JmxConnection implements AutoCloseable {
+
+  private static final long MEGABYTE = 1024L * 1024L;
+  private final JMXConnector jmxConnector;
+
+  JmxConnection(JMXConnector jmxConnector) {
+    this.jmxConnector = jmxConnector;
+  }
+
+  /**
+   * Get a MBean from a remote JMX server.
+   * @throws IllegalStateException if a valid
+   * connection to remote server cannot be created, for instance because the connection to has
+   * not yet been established (with {@link JMXConnector#connect()}), or it has been closed/broken.
+   */
+  public <M> M getMBean(String mBeanName, Class<M> mBeanInterfaceClass) {
+    try {
+      MBeanServerConnection connection = jmxConnector.getMBeanServerConnection();
+      if (mBeanName.startsWith("java.lang")) {
+        return ManagementFactory.newPlatformMXBeanProxy(connection, mBeanName, mBeanInterfaceClass);
+      }
+      return JMX.newMBeanProxy(connection, new ObjectName(mBeanName), mBeanInterfaceClass);
+    } catch (Exception e) {
+      throw new IllegalStateException("Fail to connect to MBean " + mBeanName, e);
+    }
+  }
+
+  public SortedMap<String, Object> getSystemState() {
+    SortedMap<String, Object> props = new TreeMap<>();
+    MemoryMXBean memory = getMBean(ManagementFactory.MEMORY_MXBEAN_NAME, MemoryMXBean.class);
+    props.put("Heap Committed (MB)", toMegaBytes(memory.getHeapMemoryUsage().getCommitted()));
```

Comment:
you should store `memory.getHeapMemoryUsage()` and `memory.getNonHeapMemoryUsage()` into a variable to be sure you get consistent values + that would make sure there is no performance issue if these methods do some IOs